Sort and indent packages/package groups in rpm-ostree status

Mershl opened the pull request:
Created this PR as a first step towards fixing #1613.

Fixed issues:
- By using the existing `print_packages()` for RemovedBasePackages and ReplacedBasePackages multiple package groups are no longer started in the same line if the next package/package group would not fit into the same starting line.
- `print_packages()` creates a `packages_sorted` array, but never actually sorts the package list.
- RemovedBasePackages and ReplacedBasePackages do no longer use the comma separator. Package groups are now either nice and short (in which case it's okay to print two in the same line), or new package groups start in a new line - which obsoletes the comma separator.

Open points:
- I removed the `rpmostreecxx::maybe_shell_quote (pkg)` call in `print_packages()` as it introduced quotes for package groups (which have spaces in it). I've not found usecases where the call is necessary. **This absolutly needs a review.**
- Up for discussion: Separate every RemovedBasePackages and ReplacedBasePackages package group with a newline?
- Up for discussion: See the extreme-cases below. The PR's implementation ensures that these extreme-cases will always be printed in their own line - thus the user can resize the terminal and eventually see it in a single line. Implementing an artificial line break in `print_packages()` would result in a nicely indented block (all lines starting behind the colon of the first column), but also hinder the layout on resize.
- `initramfs::test::test_initramfs_overlay` currently fails. I'll look into that tomorrow.

The following screenshots show two extreme cases with mesa and gnome-shell overrides where a single package group exceeds a single line. With that being said, it's much nicer than before, especially in the non-extreme cases and it formats nicely when the user grows the terminal horizontally.

## `rpm-ostree-2021.12-2.fc35.x86_64`
![image](https://user-images.githubusercontent.com/520251/140232589-5cd2f934-d0dc-4012-9380-e1b653ac6964.png)

## `rpm-ostree-main + sort_and_indent_rpmostree_status`
![image](https://user-images.githubusercontent.com/520251/140232627-c2f127f9-6d9a-40d8-a647-3acd449572c0.png)

lucab commented:
/ok-to-test

cgwalters commented on src/app/rpmostree-builtin-status.cxx:
```diff
@@ -864,10 +861,15 @@ print_one_deployment (RPMOSTreeSysroot *sysroot_proxy,
           g_string_append_printf (str, " %s", evr);
+
+          g_ptr_array_add (active_removals_grouped, g_strdup (str->str));
+          g_string_erase (str, 0, -1);
```
This call deletes all the data, right?  So I'm a bit confused...why are we keeping around `str` here at all?

Mershl replied:
Consistency with the current code base. Implementations around this allocate a buffer outside of the loop, then in the loop fill, use the buffer and erase it for the next loop. Proposal: Move autoptr of the buffer into the loop scope to make it's intention clear and avoid misuse from outer scope + rename `str` to `buf` (seen this in other places of the codebase). What do you think?

Mershl replied:
I'll try to clean up the string buffer handling a little bit with my next commit - the handling for ReplacedBasePackages is not "trivial". One has to see that the positive if branch ends with a truncate to size 0. That's why the buffer can be reused without an erase in between the two loops.

cgwalters commented:
Hi, thank you for the patch!

Can you squash this into one commit, and follow this for commit message style: https://chris.beams.io/posts/git-commit/

> I removed the rpmostreecxx::maybe_shell_quote (pkg) call in print_packages() as it introduced quotes for package groups (which have spaces in it). I've not found usecases where the call is necessary. This absolutly needs a review.

I think the rationale there is it allows copy-pasting the output back into a bash command.
I don't personally have a strong opinion on this.  The logic here is...old.  Dates to at least https://github.com/coreos/rpm-ostree/commit/4c2fab8f5e371c1116a13cdffa86690df8c64b6c and maybe 051492b3e6ba1d1c8eee9460fdcb3633ef0b781f

Now, if removing it seems important, that seems like exactly the kind of thing that's best done as a separate commit.

cgwalters commented:
>  follow this for commit message style: https://chris.beams.io/posts/git-commit/

> The logic here is...old. Dates to at least 4c2fab8 and maybe 051492b

To combine these two things - the process of reviewing this PR involved digging up and looking at 4 year old changes.  In the future, it might be someone besides you or me that needs to find *this* commit and understand more about *why* the changes were made.

Now you did include a lot of good stuff in the PR description *but* I don't want to totally rely on Github being around forever.  Perhaps just to start:


> status: Ensure lines are wrapped consistently
>
> Rework how we print `RemovedBasePackages` and `ReplacedBasePackages` to consistently
line wrap.

or so?

Mershl commented:
Squashed commits.
Split removal of `maybe_shell_quote` into separate commit.
Avoid existing string buffer reusage. This in turn avoids the necessary erase at the end of a loop.
-> Ready for Review.

Up for discussion: Separate every RemovedBasePackages and ReplacedBasePackages package group with a newline?

Unrelated to this change: https://github.com/Mershl/rpm-ostree/blob/0617b57f615c5c57d1a0fcd5d1c5d40eacb35108/src/app/rpmostree-builtin-status.cxx#L886
This string buffer's scope can be reduced to avoid confusion/misuse. The later GLNX_HASH_TABLE_FOREACH_KV was reusing it before, under the assumption that the previous users last iteration ended with a truncation to 0 (which in my tests always happens). I've introduced a separate local string buffer for the FOREACH_KV to make the code more readable and less error prone.

jlebon commented on src/app/rpmostree-builtin-status.cxx:
```diff
@@ -113,8 +113,7 @@ print_packages (const char *k, guint max_key_len,
       if (omit_pkgs != NULL && g_strv_contains (omit_pkgs, pkg))
         continue;
 
-      auto quoted = rpmostreecxx::maybe_shell_quote (pkg);
```
There are many other callers of `print_packages` which still rely on this. To answer your comment, the use case for this is that "packages" in some context is a loose term, and actually supports the same syntax dnf does, such as arbitrary provides or reldeps (e.g. this is valid: `rpm-ostree install "foo > 2.1"`).

Maybe what we could do instead is add a new `quote` gboolean parameter to `print_packages` so that way the callers can say if quotes may be necessary. (And then we could pass `FALSE` for the grouped packages).

Mershl replied:
Agreed on the additional function parameter. I'll update the PR shortly.

jlebon commented:
Hmm, instead of passing to `print_packages` an array of grouped packages, another option is to for it to be completely broken up, e.g. `["foo", "foo-ext", "1.2-3", "->", "1.2-4,"]`, etc... Then there are many more wrapping points available when outputting. Do you want to try that out? I think it'd result in a nicer output. (And yeah, we really should rename that function to just e.g. `print_array` or so.)

cgwalters reviewed (approved):
Hmm.  I think on the other hand it wouldn't look as good to wrap at `->` in most cases.  I can see that being necessary for small terminals and/or very long package names as an alternative strategy.  This problem space isn't new of course, it's basically the same problem as [hyphenation](https://developer.mozilla.org/en-US/docs/Web/CSS/hyphens).

I think this PR is enough of an improvement to merge as is - we can always circle back and do more improvements later.

Thanks for your work on this!
